cpp/include/cuvs/neighbors/ivf_pq.hpp

```diff
@@ -278,6 +278,72 @@ constexpr typename list_spec<SizeT, IdxT>::list_extents list_spec<SizeT, IdxT>::
 template <typename IdxT, typename SizeT = uint32_t>
 using list_data = ivf::list<list_spec, SizeT, IdxT>;
 
+using pq_centers_extents = std::experimental::
+  extents<uint32_t, raft::dynamic_extent, raft::dynamic_extent, raft::dynamic_extent>;
+
+template <typename IdxT>
+class index_iface {
+ public:
+  virtual ~index_iface() = default;
+
+  virtual cuvs::distance::DistanceType metric() const noexcept           = 0;
+  virtual codebook_gen codebook_kind() const noexcept                    = 0;
+  virtual uint32_t dim() const noexcept                                  = 0;
+  virtual uint32_t dim_ext() const noexcept                              = 0;
+  virtual uint32_t rot_dim() const noexcept                              = 0;
+  virtual uint32_t pq_bits() const noexcept                              = 0;
+  virtual uint32_t pq_dim() const noexcept                               = 0;
+  virtual uint32_t pq_len() const noexcept                               = 0;
+  virtual uint32_t pq_book_size() const noexcept                         = 0;
+  virtual uint32_t n_lists() const noexcept                              = 0;
+  virtual bool conservative_memory_allocation() const noexcept           = 0;
+  virtual uint32_t get_list_size_in_bytes(uint32_t label) const noexcept = 0;
+
+  virtual std::vector<std::shared_ptr<list_data<IdxT>>>& lists() noexcept             = 0;
+  virtual const std::vector<std::shared_ptr<list_data<IdxT>>>& lists() const noexcept = 0;
+
+  virtual raft::device_vector_view<uint32_t, uint32_t, raft::row_major> list_sizes() noexcept = 0;
+  virtual raft::device_vector_view<const uint32_t, uint32_t, raft::row_major> list_sizes()
+    const noexcept = 0;
+
+  virtual raft::device_vector_view<uint8_t*, uint32_t, raft::row_major> data_ptrs() noexcept = 0;
```

Comment:
Whya re all the possible type comtinations separate virtuals? Are we not able to have templates for these? This seems unnecessary to have these defined on the iface for all types at the same time.